pypgstac/pypgstac/db.py

```diff
@@ -0,0 +1,251 @@
+from psycopg_pool import ConnectionPool
+import atexit
+import logging
+from pydantic import BaseSettings
+from tenacity import retry, retry_if_exception_type, stop_after_attempt
+
+
+def dumps(data: dict) -> str:
+    """Dump dictionary as string."""
+    return orjson.dumps(data).decode()
+
+
+set_json_dumps(dumps)
+set_json_loads(orjson.loads)
+
+
+def pg_notice_handler(notice: psycopg.errors.Diagnostic) -> None:
+    """Add PG messages to logging."""
+    msg = f"{notice.severity} - {notice.message_primary}"
+    logging.info(msg)
+
+
+class Settings(BaseSettings):
+    """Base Settings for Database Connection."""
+
+    db_min_conn_size: int = 0
+    db_max_conn_size: int = 1
+    db_max_queries: int = 5
+    db_max_idle: int = 5
+    db_num_workers: int = 1
+    db_retries: int = 3
+
+    class Config:
+        """Use .env file if available."""
+
+        env_file = ".env"
+
+
+settings = Settings()
+
+
+class PgstacDB:
+    """Base class for interacting with PgStac Database."""
+
+    def __init__(
+        self,
+        dsn: Optional[str] = "",
+        pool: Optional[ConnectionPool] = None,
+        connection: Optional[Connection] = None,
+        commit_on_exit: bool = True,
+        debug: bool = False,
+    ) -> None:
+        """Initialize Database."""
+        self.dsn: str
+        if dsn is not None:
+            self.dsn = dsn
+        else:
+            self.dsn = ""
+        self.pool = pool
+        self.connection = connection
+        self.commit_on_exit = commit_on_exit
+        self.initial_version = "0.1.9"
+        self.debug = debug
+        if self.debug:
+            logging.basicConfig(level=logging.DEBUG)
+
+    def get_pool(self) -> ConnectionPool:
+        """Get Database Pool."""
+        if self.pool is None:
+            self.pool = ConnectionPool(
+                conninfo=self.dsn,
+                min_size=settings.db_min_conn_size,
+                max_size=settings.db_max_conn_size,
+                max_waiting=settings.db_max_queries,
+                max_idle=settings.db_max_idle,
+                num_workers=settings.db_num_workers,
+                kwargs={
+                    "options": "-c search_path=pgstac,public"
+                    " -c application_name=pypgstac"
+                },
+            )
+        return self.pool
+
+    def open(self) -> None:
+        """Open database pool connection."""
+        print("open")
+        self.get_pool()
+
+    def close(self) -> None:
+        """Close database pool connection."""
+        print("close")
+        if self.pool is not None:
+            self.pool.close()
+
+    def connect(self) -> Connection:
+        """Return database connection."""
+        pool = self.get_pool()
+        if self.connection is None:
+            self.connection = pool.getconn()
+            if self.debug:
+                self.connection.add_notice_handler(pg_notice_handler)
+        atexit.register(self.disconnect)
```

Comment:
This should tab in; if self.connection is not None then the DB is already connected, so another exit handler shouldn't be registered.